### PR TITLE
Auth library

### PR DESCRIPTION
**Pull request by valentinyanakiev**
Draft auth library

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,10 +53,13 @@
   },
   "dependencies": {
     "adal-node": "^0.2.1",
+    "axios": "^0.21.0",
+    "dotenv": "^8.2.0",
     "graphql": "^15.4.0",
     "graphql-request": "^3.3.0",
     "graphql-tag": "^2.11.0",
     "jsonpath": "^1.0.2",
+    "querystring": "^0.2.0",
```
Comment (atstoyanov) [inline package.json]: Not used?

File: src/get.token.ropc.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { AuthenticationClient } from './authentication/authentication.client';
+import { IToken } from './contracts/token';
+import * as dotenv from 'dotenv';
+
+const main = async () => {
+
+  dotenv.config();
+
```
Comment (atstojanov) [inline src/get.token.ropc.ts:8]: Break the process if any configuration parameter is missing. In the early version of the server we used `envalid` package to validate env config.

**Review by atstoyanov — COMMENTED** [review]
Check comments

File: src/get.token.ropc.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    console.log(res);
+  }
+
+};
+
+try {
+  main();
+} catch (error) {
+  console.error(error);
+}
```
Comment (atstoyanov) [inline src/get.token.ropc.ts]: For testing purposes this is OK, but the project will be used as library. All of this information should come from outside. For future to be moved to a test suite.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,27 @@
...
+AUTH_AAD_TENANT=[tenant (directory) ID]
+AUTH_AAD_CLIENT_APP_ID= [client (application) ID]
+AUTH_AAD_API_SCOPE= [API Scopes Requested for access]
+AUTH_AAD_CLIENT_APP_SECRET=[App Client Secret obtained from a web app registration*]
+AUTH_AAD_USER_UPN=[UPN of the user, e.g. xxx@yyy.onmicrosoft.com]
+AUTH_AAD_USER_PASSWORD=[user password]
+
+```
+
+1. run npm run get-access-token in terminal
```
Comment (atstoyanov) [inline README.md:27]: Describe the authentication.client methods and input parameters. This information is for development purposes and should be in separate readme file.

File: src/authentication/authentication.client.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    try {
+      token = await this.authenticate(params);
+    } catch (error) {
+      return error.response.data as ITokenError;
+    }
+
+    return token as IToken;
+  }
+}
```
Comment (atstoyanov) [inline src/authentication/authentication.client.ts]: Should be reexported into `index.ts`. Together with all interfaces/types that could be used outside the app. like, token/tokenerror etc.

File: src/contracts/auth.config.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export interface IAuthConfig {
```
Comment (atstoyanov) [inline src/contracts/auth.config.ts]: Do not prefix interface names. The `eslint` rule for that will be enforced.

File: src/get.token.ropc.ts (hunk 3 above)
Comment (valentinyanakiev) [reply src/get.token.ropc.ts]: It's for testing purposes only.

File: README.md (hunk 4 above)
Comment (valentinyanakiev) [reply README.md:27]: I have put them as inline comments, like normal library in 27df0236a92a4ee09355ebf06d8819a8fefd3c42. Is that OK?

File: src/get.token.ropc.ts (hunk 3 above)
Comment (valentinyanakiev) [reply src/get.token.ropc.ts]: It's for testing only.

File: src/authentication/authentication.client.ts (hunk 5 above)
Comment (valentinyanakiev) [reply src/authentication/authentication.client.ts]: Resolved in 27df0236a92a4ee09355ebf06d8819a8fefd3c42.

File: src/contracts/auth.config.ts (hunk 6 above)
Comment (valentinyanakiev) [reply src/contracts/auth.config.ts]: Resolved in 27df0236a92a4ee09355ebf06d8819a8fefd3c42.

File: package.json (hunk 1 above)
Comment (valentinyanakiev) [reply package.json]: Removed in 7c07bc9316eaa1106627849e8b2d026c6f47beaf.

**Review by atstoyanov — APPROVED** [review]

**Pull request merged**